### DISCUSSION TITLE
How to obtain the bound on the objective?

### DISCUSSION THREAD

**Discussion by abeham** [coin-or/Sonnet#19] [Q&A] [answered]
Where is the property or method that I can call to obtain the best bound given by the LP relaxation (or solving the dual) on the objective when using CBC through Sonnet? The Objective property only provides a `Value` - is that the objective of the best feasible integer or the best objective of the LP relaxation?

I'm only familiar with other APIs (e.g. `ObjBound` vs `ObjVal` in Gurobi).

**Comment by jhmgoossens** [comment]
Objective.Value of Sonnet returns the value of OsiSolverInterface::getObjValue(). For LPs, this is the primal objective value. For MIPs, this is the best feasible integer found so far.
You can reach the underlying Osi solver by using the Sonnet solver.OsiSolver property to interrogate the OsiSolver's properties in case a property is not available via Sonnet natively. 
For MIPs solved with CBC in Sonnet, you can use the CbcModel's getBestObjectiveValue() to get the bound:
```
OsiCbcSolverInterface osisolver = solver.OsiSolver as OsiCbcSolverInterface;
CbcModel cbcModel = osisolver.getModelPtr();
double best = cbcModel.getBestPossibleObjValue();
```
See also [Sonnet_CbcTests.cs line 123](https://github.com/coin-or/Sonnet/blob/master/test/SonnetTest/Sonnet_CbcTests.cs#L123).

I'm not sure there _is_ a property within OsiSolverInterface that provides the value you're looking for. Let's look closer.

Note: SonnetWrapper does the wrapping of the native C++ classes to .NET.  Therefore, any property of the native C++ class (like OsiSolverInterface::foo) first needs to be wrapped by SonnetWrapper, but this is not done exhaustively for all properties, so you might not find what you expect. But updating the wrappers is usually simple.

**Comment by jhmgoossens** [comment, marked as answer]
I’ve added a Bound property to Objective. This only works for solver OsiCbc, in which case Bound will have the value of getBestPossibleObjValue(). Bound value is the best LP objective of all nodes remaining on the search tree. As such, Bound is the other side of the optimality gap from Value. 

For now, this feature is only available in the dev-obj-bound branch, with [binaries on AppVeyor](https://ci.appveyor.com/project/coin-or/sonnet/builds/41958773/artifacts).

Is this what you’re looking for?

**Reply by abeham** [reply]
Thanks, that's very kind of you. I have tested it with the binaries from AppVeyor and it works.
